libraries/lib-string-utils/FromChars.cpp

```diff
@@ -0,0 +1,190 @@
+/*!********************************************************************
+
+ Audacity: A Digital Audio Editor
+
+ @file FromChars.cpp
+ @brief Define functions to convert numeric types to string representation.
+
+ Dmitry Vedenko
+ **********************************************************************/
+
+#include "FromChars.h"
+
+#include "3party/fast_float.h"
+
+#include <limits>
+#include <algorithm>
+
+namespace
+{
+// Converting c to unsigned here helps to eliminate one check 
+unsigned digitToInt(char c) noexcept
+{
+   return unsigned(c - '0');
+}
+
+template <typename ResultType>
+FromCharsResult FastStringToInt(
+   const char* first, const char* last, ResultType& value,
+   bool isNegative) noexcept
+{
+   using UnsignedResultType = std::make_unsigned_t<ResultType>;
+
+   const auto availableBytes = last - first;
+
+   if (availableBytes <= 0)
+      return { first, std::errc::invalid_argument };
+
+   UnsignedResultType result = digitToInt(*first);
+
+   if (result > 10)
+      return { first, std::errc::invalid_argument };
+
+   constexpr auto maxSafeDigits = std::numeric_limits<ResultType>::digits10 - 1;
+
+   const char* ptr = first;
+   const char* safeLast =
+      first + std::min<decltype(availableBytes)>(availableBytes, maxSafeDigits);
+
+   unsigned d;
+
+   // No integer overflow here
+   while (++ptr < safeLast && (d = digitToInt(*ptr)) <= 9)
+   {
+      result = result * 10 + d;
+   }
+
+   // But here live dragons
+   while (ptr < last && (d = digitToInt(*ptr++)) <= 9)
+   {
+      const UnsignedResultType maxValue =
+         std::is_signed_v<ResultType> && isNegative ?
+            static_cast<UnsignedResultType>(
+               std::numeric_limits<ResultType>::max() + 1) :
```

Comment:
Should this be
`std::is_signed_v<ResultType>
   ? static_cast<UnsignedResultType>(
      std::numeric_limits<ResultType>::max() / 2 + (isNegative ? 1 : 0))
   : std::numeric_limits<ResultType>::max()`

Reply:
Hmm, could be a bit faster, yes!

Reply:
Not just faster, but correct.  It was wrong without `/ 2`.  You want overflow of the absolute value of the signed value at a lower threshold than with unsigned.

Reply:
And you want the difference of 1 in that limit for the negative and positive cases.

Reply:
Sorry, I was wrong, because the template argument is <ResultType> not <UnsignedRestultType>

Reply:
But the original was wrong too, because the cast to UnsignedResultType should be done before adding 1 !

Reply:
> But the original was wrong too, because the cast to UnsignedResultType should be done before adding 1 !  (Maybe it works, but the small change technically avoids actual overflow in the computation of maxValue itself)

Reply:
Yes, I rewrote it correctly